### PR TITLE
#3371 Active startup project is calculated incorrectly 

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
Properly figure out startup project

### CODE REVIEW

File: src/Package/Impl/ProjectSystem/ProjectSystemServices.cs [hunk 1]
```diff
@@ -28,7 +29,11 @@ internal sealed class ProjectSystemServices : IProjectSystemServices {
             DTE dte = VsAppShell.Current.GetGlobalService<DTE>();
             if (dte.Solution.Projects.Count > 0) {
                 try {
-                    return dte.Solution?.Projects?.Cast<EnvDTE.Project>()?.First();
+                    var projects = dte.Solution?.SolutionBuild?.StartupProjects as IEnumerable;
```
Comment (AlexanderSher) [inline src/Package/Impl/ProjectSystem/ProjectSystemServices.cs]: `var projectName = dte.Solution?.SolutionBuild?.StartupProjects?.FirstOrDefault() as string;`

Comment (AlexanderSher) [inline src/Package/Impl/ProjectSystem/ProjectSystemServices.cs]: `var projectName = dte.Solution?.SolutionBuild?.StartupProjects?.Cast<string>().FirstOrDefault();`

File: src/Package/Impl/ProjectSystem/ProjectSystemServices.cs [hunk 2]
```diff
@@ -28,7 +29,11 @@ internal sealed class ProjectSystemServices : IProjectSystemServices {
...
+                    var e = projects?.GetEnumerator();
+                    e?.MoveNext();
+                    var projectName = e?.Current as string;
+                    return !string.IsNullOrEmpty(projectName) ? dte.Solution.Projects.Item(projectName) : null;
```
Comment (AlexanderSher) [inline src/Package/Impl/ProjectSystem/ProjectSystemServices.cs]: In old code, both `Solution` and `Projects` properties had null checks

Comment (MikhailArkhipov) [inline src/Package/Impl/ProjectSystem/ProjectSystemServices.cs]: if it DID find project name, they are not null. But we can be paranoid :-)

File: src/Package/Impl/ProjectSystem/ProjectSystemServices.cs (hunk 1 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/ProjectSystem/ProjectSystemServices.cs]: StartupProjects is `object`.

Comment (MikhailArkhipov) [inline src/Package/Impl/ProjectSystem/ProjectSystemServices.cs]: Samples only allow `foreach(string s in startupProjects)`

Comment (int19h) [inline src/Package/Impl/ProjectSystem/ProjectSystemServices.cs]: If it's actually `IEnumerable` (which your change implies by casting it to one), then you can do `Cast` to get a typed enumerable out of it.

Comment (AlexanderSher) [inline src/Package/Impl/ProjectSystem/ProjectSystemServices.cs]: From https://msdn.microsoft.com/en-us/library/envdte.solutionbuild.startupprojects.aspx:
```cs
        SolutionBuild sb = dte.Solution.SolutionBuild;
        SolutionConfiguration sc = sb.ActiveConfiguration;
        vsBuildState vsBS;
        string msg = "Return relative path to startup projects: ";
        foreach (String s in (Array)sb.StartupProjects) {
            msg += "\n   " + s;
        }
```
So it is an array. In this case,
```cs
var projectName = (dte.Solution?.SolutionBuild?.StartupProjects as IEnumerable)?.Cast<string>().FirstOrDefault();
```

Comment (MikhailArkhipov) [inline src/Package/Impl/ProjectSystem/ProjectSystemServices.cs]: OK. As long as it works on native COM `IEnum***`

Comment (MikhailArkhipov) [inline src/Package/Impl/ProjectSystem/ProjectSystemServices.cs]: Speaking of casting to `Array` I wouldn't be so sure. Implementation dependent.

**Comment by AlexanderSher** [comment]
👍

**Pull request merged**